feat: bump up vllm version to 0.11.2 & remove python 3.9 support

csy1204 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kserve/kserve/blob/master/docs/DEVELOPER_GUIDE.md
2. Before raising a PR, please run `make precommit` to check the code style.
3. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
4. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4850

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
  ## Major Changes

  ### Dependencies
  - **vLLM**: Upgraded from `0.9.2` to `0.11.2`
  - **PyTorch**: Upgraded from `2.7.0` to `2.9.0`
  - **CUDA**: Upgraded from `12.8.1` to `12.9.1`
  - **Python**: Minimum version increased from `3.9` to `3.10` (supports up to `3.13`)
  - **Triton**: Upgraded from `>=3.0.0` to `>=3.2.0`
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

csy1204 committed:
refactor: Update VLLMModel to improve tool and reasoning parser validation, enhance error handling, and streamline chat template processing

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
refactor: Simplify argument handling in VLLMModel by removing getattr calls for default values

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
refactor: Simplify argument handling in VLLMModel by removing getattr calls for default values

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
chore: Update VLLM version to 0.11.2 in Dockerfiles and setup script

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
refactor: Update import path for FlexibleArgumentParser and streamline AsyncLLM handling in VLLM

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
refactor: Update reasoning parser handling in VLLMModel to use structured outputs configuration

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
refactor: Remove guided_decoding_backend parameter from multiple test functions in VLLM tests

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
refactor: Update import path for FlexibleArgumentParser in server tests

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
Merge branch 'master' into feat/vllm-0.11.2

csy1204 committed:
update pyproject.toml & dockerfiles

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
revert ubuntu version in dockerfile

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
update lmcache installation in dockerfile

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
add nvidia-nvshmem-cu12 ignore list

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 commented on python/pyproject.toml:
```diff
@@ -32,6 +32,7 @@ ignore-packages = [
   "nvidia-nccl-cu12", 
   "nvidia-nvjitlink-cu12", 
   "nvidia-nvtx-cu12", 
+  "nvidia-nvshmem-cu12",
```
fix pip license error: `ValueError: License 'Other/Proprietary License' is not in the allowed list found for package nvidia-nvshmem-cu12`

csy1204 commented on python/huggingface_server.Dockerfile:
```diff
@@ -96,18 +94,12 @@ RUN --mount=type=cache,target=/root/.cache/pip pip install lmcache==${LMCACHE_VE
-    cd flashinfer && \
-    python3 -m flashinfer.aot && \
-    pip install --no-build-isolation . && \
-    cd .. && rm -rf flashinfer; \
-  fi
+    pip install flashinfer-cubin==${FLASHINFER_VERSION} && \
+    pip install flashinfer-jit-cache==${FLASHINFER_VERSION} \
+        --extra-index-url https://flashinfer.ai/whl/cu$(echo ${CUDA_VERSION} | cut -d. -f1,2 | tr -d '.') && \
+    flashinfer show-config
```
ref: https://github.com/vllm-project/vllm/blob/d44a63c6d6e1a545aff270b3b85cf231ef779dab/docker/Dockerfile#L343-L349

csy1204 committed:
add nvidia-cutlass-dsl in ignore list

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 commented:
@sivanantha321  Could you please review this PR and trigger the CI tests when you have time? The build and tests are passing locally on my side.

sivanantha321 commented:
> @sivanantha321 Could you please review this PR and trigger the CI tests when you have time? The build and tests are passing locally on my side.

triggered.

sivanantha321 commented:
@csy1204 Can you run `make precommit` and commit the changes ?

Commit:
Update torchvision version in Dockerfile to 0.24.0

Signed-off-by: 조상연[플레이스 AI] <sang-yeon.cho@navercorp.com>

Commit:
Update orjson version in uv.lock files and adjust dependency markers for platform compatibility in kserve and aiffairness projects

Signed-off-by: 조상연[플레이스 AI] <sang-yeon.cho@navercorp.com>

Commit:
Update Intel Extension for PyTorch version in Dockerfile to 2.8.0

Signed-off-by: 조상연[플레이스 AI] <sang-yeon.cho@navercorp.com>

Commit:
Update PyTorch and torchvision versions in Makefile to 2.9.0 and 0.24.0 respectively

Signed-off-by: 조상연[플레이스 AI] <sang-yeon.cho@navercorp.com>

csy1204 commented:
@sivanantha321 All set — ran make precommit and fixed the test failures. Thanks!

csy1204 committed:
Merge branch 'master' into feat/vllm-0.11.2

csy1204 committed:
fix test error

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
fix torchvision wheel version error

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
Merge branch 'master' into feat/vllm-0.11.2

csy1204 committed:
Merge branch 'master' into feat/vllm-0.11.2

csy1204 committed:
add backend huggingface arg to token classification test

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
fix: resolve vllm CPU build dependency conflict for huggingfaceserver

   The kserve[llm] dependency on vllm==0.11.2 requires torchvision==0.24.0,
   but CPU builds need torch==2.8.0+cpu with torchvision==0.23.0+cpu.

   Changes:
   - Remove direct torch, triton, vllm deps from huggingfaceserver pyproject.toml
   - Update install_cpu_dependencies to temporarily replace kserve[llm] with kserve
   - Modify workflow to run setup_vllm.sh last to preserve torch CPU version
   - Use uv pip in setup_vllm.sh for consistency

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
fix: resovle huggingfaceserver test errors

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 committed:
env: remove python 3.9 env

Signed-off-by: csy1204 <josang1204@gmail.com>

csy1204 commented:
/rerun-all

<details>
<summary><b>python-test.yml huggingfaceserver-cpu test result (100%)<b></summary>

```
  /home/csy/projects/csy-kserve/python/huggingfaceserver/.venv/lib/python3.12/site-packages/kserve/protocol/rest/openai/openai_chat_adapter_model.py:76: DeprecationWarning: max_tokens is deprecated in favor of the max_completion_tokens field
    max_tokens=request.max_tokens,

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform linux, python 3.12.11-final-0 ----------
Name                                                   Stmts   Miss  Cover
--------------------------------------------------------------------------
huggingfaceserver/__init__.py                              9      0   100%
huggingfaceserver/__main__.py                            113    113     0%
huggingfaceserver/async_generate_stream.py                20     20     0%
huggingfaceserver/encoder_model.py                       285     85    70%
huggingfaceserver/generative_model.py                    269     31    88%
huggingfaceserver/request_logger.py                       20     15    25%
huggingfaceserver/stop_sequence_stopping_criteria.py      18      1    94%
huggingfaceserver/task.py                                 41      5    88%
huggingfaceserver/time_series_model.py                    90     10    89%
huggingfaceserver/utils.py                                70     30    57%
--------------------------------------------------------------------------
TOTAL                                                    935    310    67%

===================================================== 34 passed, 51 deselected, 5 warnings in 37.43s ======================================================
```
</details>

csy1204 committed:
Merge branch 'master' into feat/vllm-0.11.2

Jooho committed:
Merge branch 'master' into feat/vllm-0.11.2

sivanantha321 committed:
Merge branch 'master' into feat/vllm-0.11.2

sivanantha321 commented:
@csy1204 looks like the test is failling with module not found error.
`E   ModuleNotFoundError: No module named 'ray'`

csy1204 committed:
Merge branch 'master' into feat/vllm-0.11.2

Commit:
fix: ignore gpu test for hf test

Signed-off-by: 조상연[플레이스 AI] <sang-yeon.cho@navercorp.com>

csy1204 commented:
> E   ModuleNotFoundError: No module named 'ray'

@sivanantha321 @Jooho Yes, The failure is from `ModuleNotFoundError: ray`. Ray is only needed for the [CUDA/GPU path](https://github.com/vllm-project/vllm/blob/275de34170654274616082721348b7edd9741d32/requirements/cuda.txt#L7) in vLLM, and we don’t install it in the CPU test environment because adding it pulls a **[torch/vLLM version mismatch](https://github.com/vllm-project/vllm/blob/275de34170654274616082721348b7edd9741d32/requirements/cpu.txt#L10-L12)**. The health check test isn’t part of our functional test coverage here, so I’ll ignore it in this context for now.

Jooho commented on .github/workflows/python-test.yml:
```diff
@@ -280,7 +280,7 @@ jobs:
           bash tests/setup_vllm.sh
           source /mnt/python/huggingfaceserver-cpu-venv/bin/activate
-          /mnt/python/huggingfaceserver-cpu-venv/bin/python -m pytest --cov=huggingfaceserver -vv -k 'not test_vllm'
+          /mnt/python/huggingfaceserver-cpu-venv/bin/python -m pytest --cov=huggingfaceserver -vv -k 'not test_vllm' --ignore=test_health_check.py
```
Since our CI environment can’t run GPU tests and only uses the CPU HuggingFace server, we still need to keep this test instead of skipping it — especially because we also have a GPU HuggingFace server. Keeping the test ensures that both CPU and GPU(multinode) work correctly.

Jooho replied:
Can you please restore this part?

csy1204 replied:
@Jooho Sure, Thanks for your help!

Jooho reviewed (commented):
I’m testing some changes in the test code to make sure this part works correctly without skipping the test. please give me some time.

Jooho committed:
Merge branch 'master' into feat/vllm-0.11.2

csy1204 commented on .github/workflows/python-test.yml:
```diff
@@ -259,30 +259,28 @@ jobs:
           bash tests/setup_vllm.sh
           source /mnt/python/huggingfaceserver-cpu-venv/bin/activate
-          /mnt/python/huggingfaceserver-cpu-venv/bin/python -m pytest --cov=huggingfaceserver -vv -k 'not test_vllm'
+          /mnt/python/huggingfaceserver-cpu-venv/bin/python -m pytest --cov=huggingfaceserver -vv -k 'not test_vllm' --ignore=test_health_check.py
```
```suggestion
          /mnt/python/huggingfaceserver-cpu-venv/bin/python -m pytest --cov=huggingfaceserver -vv -k 'not test_vllm'
```

csy1204 committed:
Update .github/workflows/python-test.yml

Signed-off-by: Sangyeon Cho <josang1204@gmail.com>

spolti commented on python/huggingfaceserver/Makefile:
```diff
@@ -1,30 +1,31 @@
-TORCH_VERSION ?= 2.7.0
-TORCHVISION_VERSION ?= 0.22.0
+TORCH_VERSION ?= 2.8.0
+TORCHVISION_VERSION ?= 0.23.0
```
Shouldn't this be on the same version than the dockerfiles?

csy1204 replied:
@spolti These versions were initially aligned with vLLM CPU dependencies based on the `install_cpu_dependencies` flow. However, we no longer use this path, and this looks like leftover config from earlier debugging. I agree we should remove it.

```
install_cpu_dependencies:
	uv add --active \
		"torch~=${TORCH_VERSION}" \
		"torchaudio~=${TORCH_VERSION}" \
		"torchvision~=${TORCHVISION_VERSION}" \
		--index ${TORCH_EXTRA_INDEX_URL} \
		--index-strategy unsafe-best-match
```

spolti commented on python/huggingfaceserver/Makefile:
```diff
@@ -1,30 +1,31 @@
+# 2. Then run setup_vllm.sh LAST to build vLLM from source with torch CPU
+#
+# The key issue: kserve[llm] depends on vllm==0.11.2 which requires torchvision==0.24.0,
+# but torch==2.8.0+cpu requires torchvision==0.23.0+cpu. To avoid this conflict:
```
does it help if you first install everything as it is and later do a manuall `uv add torchvision==0.23.0+cpu`

csy1204 replied:
@spolti 

I already tried that, but the issue is that uv add validates and resolves compatibility with the previously installed set. So after installing everything “as is,” a manual uv add torchvision==0.23.0+cpu triggers a re-resolution that conflicts with the vLLM GPU compatibility/dependency constraints(torchvision==0.24.0).

csy1204 commented on python/huggingfaceserver/Makefile:
```diff
@@ -1,30 +1,31 @@
-TORCHVISION_VERSION ?= 0.22.0
+TORCH_VERSION ?= 2.8.0
+TORCHVISION_VERSION ?= 0.23.0
+IPEX_VERSION ?= 2.8.0
```
```suggestion
```

csy1204 committed:
Update python/huggingfaceserver/Makefile

Signed-off-by: Sangyeon Cho <josang1204@gmail.com>

spolti commented:
Hi @csy1204, thanks for confirming.

spolti commented:
/lgtm

Jooho commented:
https://github.com/kserve/kserve/pull/4892 will fix the ci failure. let's wait for this.

Jooho committed:
Merge branch 'master' into feat/vllm-0.11.2

vivekk16 commented:
/rerun-all

csy1204 commented on .github/workflows/e2e-test-llmisvc.yaml:
```diff
@@ -39,7 +39,7 @@ jobs:
         uses: actions/setup-python@v5
         with:
-          python-version: "3.9"
+          python-version: "3.10"
```
```suggestion
          python-version: "3.9"
```

csy1204 committed:
Update .github/workflows/e2e-test-llmisvc.yaml

Signed-off-by: Sangyeon Cho <josang1204@gmail.com>

Jooho committed:
Merge branch 'master' into feat/vllm-0.11.2

vivekk16 commented:
/rerun-all

sivanantha321 committed:
Merge branch 'master' into feat/vllm-0.11.2

sivanantha321 commented:
precommit check failure is not related to this PR changes.

vivekk16 commented:
@spolti the precommit failure is consistent across the PRs

csy1204 committed:
Merge branch 'master' into feat/vllm-0.11.2

csy1204 commented:
@vivekk16 @sivanantha321 @spolti @Jooho 
https://github.com/kserve/kserve/pull/4897 The CI failure has been addressed.
Could you please rerun the CI? If everything looks good, we can go ahead and merge.
